src/Raven.Server/Web/System/SetupHandler.cs

```diff
@@ -0,0 +1,383 @@
+        }
+
+        [RavenAction("/setup/generate", "POST", AuthorizationStatus.UnauthenticatedClients)]
+        public async Task Generate()
+        {
+            AssertOnlyInSetupMode();
+
+            ServerStore.EnsureNotPassive();
+
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext ctx))
+            {
+                var operationId = GetLongQueryString("operationId", false);
+                if (operationId.HasValue == false)
+                    operationId = ServerStore.Operations.GetNextOperationId();
+
+                var stream = TryGetRequestFromStream("Options") ?? RequestBodyStream();
+
+                var certificateJson = ctx.ReadForDisk(stream, "certificate-generation");
+
+                var certificate = JsonDeserializationServer.CertificateDefinition(certificateJson);
+
+                byte[] pfx = null;
+                await
+                    ServerStore.Operations.AddOperation(
+                        null,
+                        "Generate certificate: " + certificate.Name,
+                        Documents.Operations.Operations.OperationType.CertificateGeneration,
+                        async onProgress =>
+                        {
+                            pfx = await GenerateCertificateInternal(certificate);
+
+                            return ClientCertificateGenerationResult.Instance;
+                        },
+                        operationId.Value);
+
+                var contentDisposition = "attachment; filename=" + Uri.EscapeDataString(certificate.Name) + ".pfx";
+                HttpContext.Response.Headers["Content-Disposition"] = contentDisposition;
+                HttpContext.Response.ContentType = "binary/octet-stream";
+
+                HttpContext.Response.StatusCode = (int)HttpStatusCode.Created;
+                HttpContext.Response.Body.Write(pfx, 0, pfx.Length);
+            }
+        }
+
+        [RavenAction("/setup/finish", "POST", AuthorizationStatus.UnauthenticatedClients)]
+        public Task SetupFinish()
+        {
+            AssertOnlyInSetupMode();
+
+            Task.Run(async () =>
+            {
+                // we want to give the studio enough time to actually
+                // get a valid response from the server before we reset
+                await Task.Delay(250);
+
+                Program.ResetServerMre.Set();
+                Program.ShutdownServerMre.Set();
+            });
+
+            return Task.CompletedTask;
+        }
+
+        [RavenAction("/admin/setup/letsencrypt/force-renew", "POST", AuthorizationStatus.ClusterAdmin)]
+        public Task ForceRenew()
+        {
+            using (ServerStore.ContextPool.AllocateOperationContext(out JsonOperationContext context))
+            using (var setupInfoJson = context.ReadForDisk(RequestBodyStream(), "setup-lets-encrypt"))
+            {
+                var setupInfo = JsonDeserializationServer.SecuredSetupInfo(setupInfoJson);
```

Comment:
Shouldn't we already have all the details we need without getting them from the outside at this point?